source/grid/grid_generator.cc

```diff
@@ -1910,6 +1910,313 @@ namespace GridGenerator
+        // 2. Finds the location of the vertex marked with 'O' and uses that to
+        //    calculate the shift vector for aligning cylinder_tria with
+        //    tria_without_cylinder.
+        std::set<Triangulation<2>::active_cell_iterator> cells_to_remove;
+        Tensor<1, 2> cylinder_triangulation_offset;
+        for (const auto &cell : bulk_tria.active_cell_iterators())
+          {
+            if ((cell->center() - Point<2>(0.2, 0.2)).norm() < 0.15)
+              cells_to_remove.insert(cell);
+
+            if (cylinder_triangulation_offset == Tensor<1, 2>())
+              {
+                for (unsigned int vertex_n = 0;
+                     vertex_n < GeometryInfo<2>::vertices_per_cell;
+                     ++vertex_n)
+                  if (cell->vertex(vertex_n) == Point<2>())
+                    {
+                      // cylinder_tria is centered at zero, so we need to
+                      // shift it up and to the right by two cells:
+                      cylinder_triangulation_offset =
+                        2.0 * (cell->vertex(3) - Point<2>());
+                      break;
+                    }
+              }
+          }
+        Triangulation<2> tria_without_cylinder;
+        GridGenerator::create_triangulation_with_removed_cells(
+          bulk_tria, cells_to_remove, tria_without_cylinder);
+
+        // set up the cylinder triangulation
+        Triangulation<2> cylinder_tria;
+        GridGenerator::hyper_cube_with_cylindrical_hole(cylinder_tria,
+                                                        0.05,
+                                                        0.41 / 4.0);
+        GridTools::shift(cylinder_triangulation_offset, cylinder_tria);
+        // Mark the cylinder material ids as 2 to distinguish them from the bulk
+        // cells (these are read again about 40 lines down)
+        for (const auto &cell : cylinder_tria.active_cell_iterators())
+          cell->set_material_id(2);
+
+        auto minimal_line_length = [](const Triangulation<2> &t) -> double {
+          double min_line_length = std::numeric_limits<double>::max();
+          for (const auto &cell : t.active_cell_iterators())
+            for (unsigned int line_n = 0;
+                 line_n < GeometryInfo<2>::lines_per_cell;
+                 ++line_n)
+              min_line_length = std::min(min_line_length,
+                                         (cell->line(line_n)->vertex(0) -
+                                          cell->line(line_n)->vertex(1))
+                                           .norm());
+          return min_line_length;
+        };
+
+        // the cylindrical triangulation might not match the Cartesian grid: as
+        // a result the vertices might not be lined up. Get around this by
+        // deleting duplicated vertices with a very low numerical tolerance.
+        const double tolerance =
+          std::min(minimal_line_length(tria_without_cylinder),
```

Comment:
since you hard-code dimension, don't you know which tolerance works in advance?

Reply:
Yes, and I hard-code other tolerances in other places. Its not consistent. Would you suggest doing all one or the other?

Reply:
I would probably hardcode it here as well to make code shorter.